File: typescript/src/VaasErrors.ts
import { ProblemDetails } from "./messages/ProblemDetails";

/** Vaas authentication failed. */
export class VaasAuthenticationError extends Error {
  constructor(message?: string) {
    super(message ?? "Vaas authentication failed");
  }
}

/** Vaas invalid state error.
 * @description These are coding errors and be prevented by the developer.
 */
export class VaasInvalidStateError extends Error {
  constructor(message: string) {
    super(message);
  }
}

/** Vaas timeout. */
export class VaasTimeoutError extends Error {
  constructor() {
    super("Timeout");
  }
}

/** Vaas server error.
 * @description The server encountered an internal error.
 */
export class VaasServerError extends Error {
  constructor(message: string | ProblemDetails) {
    if (message instanceof ProblemDetails) {
      super(message.detail);
    } else {
      super(message);
    }
  }
}

/** Vaas client error.
 * @description The request is malformed or cannot be completed.
 */
export class VaasClientError extends Error {
  public problemDetails?: ProblemDetails;

  constructor(message: string | ProblemDetails) {
    if (message instanceof ProblemDetails) {
      super(message.detail);
      this.problemDetails = message;
    } else {
      super(message);
    }
  }
}

File: python/src/vaas/vaas_errors.py
class VaasAuthenticationError(BaseException):
    """Authentication Error"""

class VaasClientError(BaseException):
    """Client Error"""
    def __init__(self, message = None) -> None:
        if message is None:
            super().__init__("Client Error")
        else:
            super().__init__(message)

class VaasServerError(BaseException):
    """Server Error"""
    def __init__(self, message = None) -> None:
        if message is None:
            super().__init__("Server Error")
        else:
            super().__init__(message)
